// appium-ios-driver/build/lib/commands/timeout.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var commands = {},
    helpers = {},
    extensions = {};

commands.timeouts = function callee$0$0(type, duration) {
  var ms;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        ms = parseInt(duration, 10);
        context$1$0.t0 = type;
        context$1$0.next = context$1$0.t0 === 'command' ? 4 : context$1$0.t0 === 'implicit' ? 6 : context$1$0.t0 === 'page load' ? 9 : context$1$0.t0 === 'script' ? 11 : 14;
        break;

      case 4:
        this.newCommandTimeout(ms);
        return context$1$0.abrupt('break', 15);

      case 6:
        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(this.implicitWait(ms));

      case 8:
        return context$1$0.abrupt('break', 15);

      case 9:
        this.pageLoadTimeout(ms);
        return context$1$0.abrupt('break', 15);

      case 11:
        context$1$0.next = 13;
        return _regeneratorRuntime.awrap(this.asyncScriptTimeout(ms));

      case 13:
        return context$1$0.abrupt('break', 15);

      case 14:
        throw new Error('Invalid timeout \'' + type + '\'');

      case 15:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.pageLoadTimeout = function (ms) {
  ms = parseInt(ms, 10);
  this.pageLoadMs = ms;
  _logger2['default'].debug('Set page load timeout to ' + ms + 'ms');
};

commands.asyncScriptTimeout = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        this.asyncWaitMs = ms;
        _logger2['default'].debug('Set async script timeout to ' + ms + 'ms');

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.pageLoadTimeout = function (ms) {
  this.pageLoadMs = ms;

  if (this.remote) {
    this.remote.pageLoadMs = ms;
  }
  _logger2['default'].debug('Set page load timeout to ' + ms + 'ms');
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports['default'] = extensions;
